### PR TITLE
4.x: Helidon Webclient (4.0.10) is not routing the requests through proxy configured using Proxy Builder. #9022

### PR DESCRIPTION
**Pull request by jbescos**
### Description
https://github.com/helidon-io/helidon/issues/9022

It will do HTTP CONNECT if any of these is true:
- You explicitly set in a flag.
- It is TLS
- It uses proxy authentication

### Documentation
None

### CODE REVIEW

**Review by klustria — DISMISSED** [review]
This LGTM already, so I'm approving. Also let's run this through @tomas-langer if he has any further comments

File: webclient/api/src/main/java/io/helidon/webclient/api/Proxy.java [hunk 1]
```diff
@@ -667,6 +677,19 @@ public Builder type(ProxyType type) {
             return this;
         }
 
+        /**
+         * Forces HTTP CONNECT with the proxy server.
+         * Otherwise it will not execute HTTP CONNECT for HTTP protocol.
```
Comment (tomas-langer) [inline webclient/api/src/main/java/io/helidon/webclient/api/Proxy.java:682]: This is not exact. Please reword it to make it clear that we talk about plain vs. TLS.
Both are HTTP protocol.

File: webclient/api/src/main/java/io/helidon/webclient/api/Proxy.java [hunk 2]
```diff
@@ -469,26 +472,26 @@ private static Socket connectToProxy(WebClient webClient,
...
-            request.header(HeaderValues.CONNECTION_KEEP_ALIVE)
-                .header(PROXY_CONNECTION);
-        }
-        proxy.proxyAuthHeader.ifPresent(request::header);
-        // we cannot close the response, as that would close the connection
-        HttpClientResponse response = request.request();
-        if (response.status().family() != Status.Family.SUCCESSFUL) {
-            response.close();
-            throw new IllegalStateException("Proxy sent wrong HTTP response code: " + response.status());
+        if (proxy.forceHttpConnect || tls || proxy.username.isPresent()) {
```
Comment (tomas-langer) [inline webclient/api/src/main/java/io/helidon/webclient/api/Proxy.java:475]: Proxy-Connection header should be set even when not using CONNECT (at least that is what curl does, and I would think that that is correct)

**Review by tomas-langer — CHANGES_REQUESTED** [review]

File: webclient/api/src/main/java/io/helidon/webclient/api/Proxy.java (hunk 2 above)
Comment (jbescos) [reply webclient/api/src/main/java/io/helidon/webclient/api/Proxy.java:475]: From the code we have, it seems currently we only set `Proxy-Connection` in the first request to the Proxy. This request may not happen with this new changes. Should we send always `Proxy-Connection` to every request when using proxy?.

From the spec, `Proxy-Connection` is a redundant header that has the same effect than `Connection`. So there is no point on setting it.

File: webclient/api/src/main/java/io/helidon/webclient/api/Proxy.java (hunk 1 above)
Comment (jbescos) [reply webclient/api/src/main/java/io/helidon/webclient/api/Proxy.java]: Done

File: webclient/api/src/main/java/io/helidon/webclient/api/Proxy.java (hunk 2 above)
Comment (klustria) [reply webclient/api/src/main/java/io/helidon/webclient/api/Proxy.java:475]: Looks like curl removed `Proxy-Connection` based on specs as you mentioned, but put it back because of a slew of bugs filed against it when it was removed: https://stackoverflow.com/questions/62722430/why-does-curl-send-a-proxy-connection-header-even-though-the-rfc-seems-to-disco. Seems that there are still lots of proxy servers out there that it s expecting this header.

Comment (jbescos) [reply webclient/api/src/main/java/io/helidon/webclient/api/Proxy.java:475]: Nice finding @klustria 

So, my understanding is that if proxy is set, we need to set `Proxy-Connection` in all the requests.

Comment (klustria) [reply webclient/api/src/main/java/io/helidon/webclient/api/Proxy.java:475]: Yeah, that's what I understand too.

Comment (jbescos) [reply webclient/api/src/main/java/io/helidon/webclient/api/Proxy.java:473]: Done. It is excluded in HTTP2 as it is said here: https://www.rfc-editor.org/rfc/rfc9113.html

[8.2.2. ](https://www.rfc-editor.org/rfc/rfc9113.html#section-8.2.2)[Connection-Specific Header Fields](https://www.rfc-editor.org/rfc/rfc9113.html#name-connection-specific-header-)

HTTP/2 does not use the Connection header field ([Section 7.6.1](https://www.rfc-editor.org/rfc/rfc9110#section-7.6.1) of [[HTTP](https://www.rfc-editor.org/rfc/rfc9113.html#RFC9110)]) to indicate connection-specific header fields; in this protocol, connection-specific metadata is conveyed by other means. An endpoint MUST NOT generate an HTTP/2 message containing connection-specific header fields. This includes the Connection header field and those listed as having connection-specific semantics in [Section 7.6.1](https://www.rfc-editor.org/rfc/rfc9110#section-7.6.1) of [[HTTP](https://www.rfc-editor.org/rfc/rfc9113.html#RFC9110)] (that is, Proxy-Connection, Keep-Alive, Transfer-Encoding, and Upgrade). Any message containing connection-specific header fields MUST be treated as [malformed](https://www.rfc-editor.org/rfc/rfc9113.html#malformed) ([Section 8.1.1](https://www.rfc-editor.org/rfc/rfc9113.html#malformed)).
The only exception to this is the TE header field, which MAY be present in an HTTP/2 request; when it is, it MUST NOT contain any value other than "trailers".An intermediary transforming an HTTP/1.x message to HTTP/2 MUST remove connection-specific header fields as discussed in [Section 7.6.1](https://www.rfc-editor.org/rfc/rfc9110#section-7.6.1) of [[HTTP](https://www.rfc-editor.org/rfc/rfc9113.html#RFC9110)], or their messages will be treated by other HTTP/2 endpoints as [malformed](https://www.rfc-editor.org/rfc/rfc9113.html#malformed) ([Section 8.1.1](https://www.rfc-editor.org/rfc/rfc9113.html#malformed)).

**Review by klustria — APPROVED** [review]
LGTM

File: webclient/http1/src/main/java/io/helidon/webclient/http1/Http1ClientRequestImpl.java [hunk 3]
```diff
@@ -179,6 +180,14 @@ protected MediaContext mediaContext() {
         return super.mediaContext();
     }
 
+    @Override
+    protected void additionalHeaders() {
+        super.additionalHeaders();
+        if (proxy() != null && proxy().type() != ProxyType.NONE) {
```
Comment (tomas-langer) [inline webclient/http1/src/main/java/io/helidon/webclient/http1/Http1ClientRequestImpl.java:186]: proxy() must never be null - this check should be useless.

**Review by tomas-langer — CHANGES_REQUESTED** [review]

File: webclient/http1/src/main/java/io/helidon/webclient/http1/Http1ClientRequestImpl.java (hunk 3 above)
Comment (jbescos) [reply webclient/http1/src/main/java/io/helidon/webclient/http1/Http1ClientRequestImpl.java]: Done

**Review by klustria — APPROVED** [review]
Looks good!

**Review by tomas-langer — APPROVED** [review]
